### DISCUSSION TITLE
What's the proper way to release processor time from within a PM program?

### DISCUSSION THREAD

**Discussion by jharrison03** [dosemu2/dosemu2#1454] [General]
Calling 2fh/1680, 28h, etc. all require dropping back into real-mode.  What's the proper way to do this from a DOS PM program?  Under Windows, a DOS program can do this through a very convoluted way through the Windows API - but how can a DOS program run under any of the common extenders accomplish this?

**Comment by stsp** [comment]
I think you already asked that,
and last time you complain was
that when DPMI switches to RM
for you to service these calls,
this is going to be too slow.
Sorry but my answer doesn't change:
"slowness" doesn't matter when
you call an idle functions.

**Comment by ecm-pushbx** [comment]
For reference, here's [lDebug's idle function](https://hg.ulukai.org/ecm/ldebug/file/bb4b40b6aa42/source/lineio.asm#l1591). In PM it first calls the Protected Mode interrupt 2Fh service 1680h, then if that fails it calls through to the 86 Mode handler for the same service number, then (only opt-in for PM) it will execute `sti` \ `hlt`

**Reply by jharrison03** [reply]
> For reference, here's [lDebug's idle function](https://hg.ulukai.org/ecm/ldebug/file/bb4b40b6aa42/source/lineio.asm#l1591). In PM it first calls the Protected Mode interrupt 2Fh service 1680h, then if that fails it calls through to the 86 Mode handler for the same service number, then (only opt-in for PM) it will execute `sti` \ `hlt`

That's the exact method I'm using now.

@stsp, I get there might not be an inefficiency with this, but this call should succeed under DPMI and it clearly doesn't - at least under whatever extender Borland C is using.  I'm trying to understand why this is the case.  And this isn't something DOSemu-related so forgive me for asking here.
